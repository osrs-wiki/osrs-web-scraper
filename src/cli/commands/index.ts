export { default as news } from "./news";
export { default as poll } from "./poll";
export { default as worlds } from "./worlds";
