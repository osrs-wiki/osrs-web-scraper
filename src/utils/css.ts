import { HTMLElement } from "node-html-parser";

/**
 * Extracts background image URL from CSS style attribute
 * Handles both single and double quotes, and HTML entity encoding
 */
export function extractBackgroundImageUrl(style: string): string | null {
  if (!style) return null;
  
  // First decode HTML entities
  const decodedStyle = style
    .replace(/&quot;/g, '"')
    .replace(/&apos;/g, "'")
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&amp;/g, '&');
  
  // Match background-image: url("...") or url('...') patterns
  const urlMatch = decodedStyle.match(/background-image\s*:\s*url\s*\(\s*['"']?([^'"')]+)['"']?\s*\)/i);
  
  if (urlMatch && urlMatch[1]) {
    return urlMatch[1];
  }
  
  return null;
}

/**
 * Extracts all background image URLs from style attributes in HTML elements
 */
export function extractBackgroundImages(elements: HTMLElement[]): string[] {
  const imageUrls: string[] = [];
  
  elements.forEach(element => {
    if (element.attributes?.style) {
      const url = extractBackgroundImageUrl(element.attributes.style);
      if (url) {
        imageUrls.push(url);
      }
    }
  });
  
  return imageUrls;
}