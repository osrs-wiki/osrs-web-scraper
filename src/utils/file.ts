import fs from "fs";
import client from "https";
import path from "path";

import { formatText } from "./text";

/**
 * Download a file from a url
 * @param url The url of the file to download
 * @param filepath The filepath of the downloaded file
 * @returns
 */
export const downloadFile = async (url: string, filepath: string) => {
  console.info(`Attempting file download: ${url}`);
  return new Promise((resolve, reject) => {
    client.get(url, (res) => {
      if (res.statusCode === 200) {
        res
          .pipe(fs.createWriteStream(filepath))
          .on("error", reject)
          .once("close", () => resolve(filepath));
        console.info(`Downloaded file ${filepath}`);
      } else {
        res.resume();
        reject(
          new Error(`Request Failed With a Status Code: ${res.statusCode}`)
        );
      }
    });
  });
};

/**
 * Get a file from a url
 * @param fileUrl The file url
 * @returns
 */
export const getFileName = (fileUrl: string) => {
  const parsed = new URL(fileUrl);
  return path.basename(parsed.pathname);
};

/**
 * Get the extension of a file from a url link.
 * @param fileUrl The file url
 * @returns
 */
export const getFileExtension = (fileUrl: string) => {
  return fileUrl.split(/[#?]/)[0].split(".").pop().trim();
};

const fileCharsReplaceMap: { [key: string]: string } = {
  ":": "-",
  "|": "-",
  "&nbsp;": "",
};

/**
 * Format a file name to ensure it is valid.
 * @param fileName The file name
 * @returns
 */
export const formatFileName = (fileName: string) => {
  const cleanedFileName = Object.keys(fileCharsReplaceMap).reduce(
    (name, find) => name.replaceAll(find, fileCharsReplaceMap[find]),
    fileName
  );
  return formatText(cleanedFileName);
};
