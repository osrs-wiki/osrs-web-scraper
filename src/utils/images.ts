import fs from "fs";
import client from "https";
import path from "path";

import { formatText } from "./text";

/**
 * Download an image from a url
 * @param url The url of the image to download
 * @param filepath The filepath of the downloaded image
 * @returns
 */
export const downloadImage = async (url: string, filepath: string) => {
  console.info(`Attempting image download: ${url}`);
  return new Promise((resolve, reject) => {
    client.get(url, (res) => {
      if (res.statusCode === 200) {
        res
          .pipe(fs.createWriteStream(filepath))
          .on("error", reject)
          .once("close", () => resolve(filepath));
        console.info(`Downloaded image ${filepath}`);
      } else {
        res.resume();
        reject(
          new Error(`Request Failed With a Status Code: ${res.statusCode}`)
        );
      }
    });
  });
};

/**
 * Get an image name from a url
 * @param imageUrl The image url
 * @returns
 */
export const getImageName = (imageUrl: string) => {
  const parsed = new URL(imageUrl);
  return path.basename(parsed.pathname);
};

/**
 * Get the extension of an image from a url link.
 * @param imageUrl The image url
 * @returns
 */
export const getImageExtension = (imageUrl: string) => {
  return imageUrl.split(/[#?]/)[0].split(".").pop().trim();
};

/**
 * Format a file name to ensure it is valid.
 * @param fileName The file name
 * @returns
 */
export const formatFileName = (fileName: string) => {
  return formatText(fileName.replaceAll(":", "-"));
};
