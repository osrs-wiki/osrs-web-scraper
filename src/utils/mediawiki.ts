import { MediaWikiBreak, MediaWikiContent } from "@osrs-wiki/mediawiki-builder";
/**
 * Trims the leading and trailing MediaWikiBreak elements from the contents.
 *
 * @param contents The contents to trim.
 * @returns The trimmed contents.
 */
export const trim = (contents: MediaWikiContent[]): MediaWikiContent[] => {
  if (!contents || contents.length === 0) {
    return contents;
  }
  while (
    contents[0] instanceof MediaWikiBreak ||
    contents[contents.length - 1] instanceof MediaWikiBreak
  ) {
    if (contents.length > 0 && contents[0] instanceof MediaWikiBreak) {
      contents.shift();
    }
    if (
      contents.length > 0 &&
      contents[contents.length - 1] instanceof MediaWikiBreak
    ) {
      contents.pop();
    }
  }
  return contents;
};

/**
 * Checks if the contents are empty.
 * @param contents The contents to check.
 * @returns True if the contents are empty, otherwise false.
 */
export const isEmpty = (
  contents: string | MediaWikiContent | MediaWikiContent[]
): boolean => {
  if (typeof contents === "string") {
    return contents.trim().length === 0;
  } else if (Array.isArray(contents)) {
    return contents.length === 0 || contents.every(isEmpty);
  } else if (contents instanceof MediaWikiContent) {
    return isEmpty(contents.children);
  }
  return false;
};

/**
 * Checks if the first child of the contents starts with the given string.
 *
 * @param contents The contents to check.
 * @param str The string to check against.
 * @returns True if the first child starts with the given string, otherwise false.
 */
export const startsWith = (
  contents: string | MediaWikiContent | MediaWikiContent[],
  str: string
): boolean => {
  if (isEmpty(contents)) {
    return false;
  }
  if (typeof contents === "string") {
    return contents.startsWith(str);
  }
  const childContents = Array.isArray(contents) ? contents[0] : contents;
  if (typeof childContents.children === "string") {
    return childContents.children.startsWith(str);
  }
  return startsWith(childContents.children, str);
};

/**
 * Recursively retrieves the first string content from the given MediaWikiContent.
 *
 * @param contents The MediaWikiContent to search through.
 * @returns The first string content found, or undefined if none is found.
 */
export const getFirstStringContent = (
  contents: MediaWikiContent
): MediaWikiContent | undefined => {
  if (typeof contents.children === "string") {
    return contents;
  } else if (contents.children instanceof MediaWikiContent) {
    return getFirstStringContent(contents.children);
  } else if (Array.isArray(contents.children) && contents.children.length > 0) {
    return getFirstStringContent(contents.children[0]);
  }
  return undefined;
};
