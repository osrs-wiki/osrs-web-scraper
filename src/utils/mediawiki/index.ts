export { default as MediaWikiBuilder } from "./builder";
export { default as MediaWikiContent } from "./content";

export * from "./contents";
