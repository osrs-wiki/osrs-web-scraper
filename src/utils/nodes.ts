import { Node } from "node-html-parser";

/**
 * Util function to prevent having to include ts-ignore every time rawTagName is referenced.
 * @param node
 * @returns
 */
export const getNodeTagName = (node: Node): string => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  return node.rawTagName;
};
