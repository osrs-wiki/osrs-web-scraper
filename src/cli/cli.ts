import { Command } from "commander";

import { news, poll, worlds } from "./commands";
import packageJson from "../../package.json";

const cli = new Command();

cli.name("OSRS Web Scraper").description("").version(packageJson.version);

cli.addCommand(news);
cli.addCommand(poll);
cli.addCommand(worlds);

export default cli;
