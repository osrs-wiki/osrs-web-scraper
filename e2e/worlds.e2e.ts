import * as fs from "fs";

import { runCLICommand, cleanupOutput } from "./utils/cli";

describe("Worlds Command E2E Tests", () => {
  // Note: Tests assume the project is already built with `npm run build`

  afterEach(async () => {
    await cleanupOutput();
  });

  it("should show help for worlds command", async () => {
    const result = await runCLICommand({
      command: "node",
      args: [
        "-r",
        "typescript-transform-paths/register",
        "./dist/src/index.js",
        "worlds",
        "--help",
      ],
    });

    expect(result.exitCode).toBe(0);
    expect(result.stdout).toContain("Scrape the OSRS world list");
  });

  it("should scrape worlds list", async () => {
    const result = await runCLICommand({
      command: "node",
      args: [
        "-r",
        "typescript-transform-paths/register",
        "./dist/src/index.js",
        "worlds",
      ],
      timeout: 180000, // 3 minutes for scraping
    });

    expect(result.exitCode).toBe(0);

    // Check that output files were created
    const outDir = "./out/worlds";
    expect(fs.existsSync(outDir)).toBe(true);

    // Check that the worlds.txt file was created
    const worldsFile = "worlds.txt";
    const filePath = `${outDir}/${worldsFile}`;
    expect(fs.existsSync(filePath)).toBe(true);

    const content = fs.readFileSync(filePath, "utf-8");

    // Verify it contains MediaWiki world list content
    expect(content).toContain("WorldLine");

    // Verify it contains a specific world line to ensure proper formatting
    expect(content).toContain(
      "{{WorldLine|302|United Kingdom|mems=yes|Trade - Members}}"
    );
  });
});
