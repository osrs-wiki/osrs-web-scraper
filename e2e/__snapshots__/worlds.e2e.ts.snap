// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Worlds Command E2E Tests should scrape worlds list: worlds-sorted.txt 1`] = `
"{{WorldLine|301|United States|mems=no|Trade - Free}}
{{WorldLine|302|United Kingdom|mems=yes|Trade - Members}}
{{WorldLine|303|Germany|mems=yes|-}}
{{WorldLine|304|Germany|mems=yes|Trouble Brewing}}
{{WorldLine|305|United States|mems=yes|Falador Party Room}}
{{WorldLine|306|United States|mems=yes|Barbarian Assault}}
{{WorldLine|307|United States|mems=yes|Wintertodt}}
{{WorldLine|308|United Kingdom|mems=no|Wilderness PK - Free}}
{{WorldLine|309|United Kingdom|mems=yes|Wintertodt}}
{{WorldLine|310|United Kingdom|mems=yes|Barbarian Assault}}
{{WorldLine|311|Germany|mems=yes|Wintertodt}}
{{WorldLine|312|Germany|mems=yes|Group Skilling}}
{{WorldLine|313|United States|mems=yes|Group Skilling}}
{{WorldLine|314|United States|mems=yes|Brimhaven Agility Arena}}
{{WorldLine|315|United States|mems=yes|Fishing Trawler}}
{{WorldLine|316|United Kingdom|mems=no|Wilderness PK - Free}}
{{WorldLine|317|United Kingdom|mems=yes|-}}
{{WorldLine|318|United Kingdom|mems=yes|Bounty Hunter World}}
{{WorldLine|320|United States|mems=yes|Soul Wars}}
{{WorldLine|321|United States|mems=yes|Sulliuscep cutting}}
{{WorldLine|322|United States|mems=yes|Clan Wars - Free-for-all}}
{{WorldLine|323|United States|mems=yes|Volcanic Mine}}
{{WorldLine|324|United States|mems=yes|Royal Titans}}
{{WorldLine|325|United Kingdom|mems=yes|Yama}}
{{WorldLine|326|United Kingdom|mems=no|LMS Casual}}
{{WorldLine|327|Germany|mems=yes|Ourania Altar}}
{{WorldLine|328|Germany|mems=yes|Royal Titans}}
{{WorldLine|329|United States|mems=yes|Tombs of Amascut}}
{{WorldLine|330|United States|mems=yes|House Party, Gilded Altar}}
{{WorldLine|331|United States|mems=yes|Tombs of Amascut}}
{{WorldLine|332|United States|mems=yes|Nex FFA}}
{{WorldLine|333|United Kingdom|mems=yes|Tombs of Amascut}}
{{WorldLine|334|United Kingdom|mems=yes|Castle Wars 1}}
{{WorldLine|335|Germany|mems=no|-}}
{{WorldLine|336|Germany|mems=yes|ToA FFA}}
{{WorldLine|337|United States|mems=yes|Nightmare of Ashihama}}
{{WorldLine|338|United States|mems=yes|ToA FFA}}
{{WorldLine|339|United States|mems=yes|Yama}}
{{WorldLine|340|United States|mems=yes|Varlamore PvM}}
{{WorldLine|341|United Kingdom|mems=yes|Tempoross}}
{{WorldLine|342|United Kingdom|mems=yes|Role-playing}}
{{WorldLine|343|Germany|mems=yes|Yama}}
{{WorldLine|344|Germany|mems=yes|Pest Control}}
{{WorldLine|345|United States|mems=deadman|Deadman}}
{{WorldLine|346|United States|mems=yes|Agility Training}}
{{WorldLine|347|United States|mems=yes|-}}
{{WorldLine|348|United States|mems=yes|PvP Arena (Legacy Duels)}}
{{WorldLine|349|United Kingdom|mems=yes|2000 skill total}}
{{WorldLine|350|United Kingdom|mems=yes|Soul Wars}}
{{WorldLine|351|Germany|mems=yes|-}}
{{WorldLine|352|Germany|mems=yes|Blast Furnace}}
{{WorldLine|353|United States|mems=yes|1250 skill total}}
{{WorldLine|354|United States|mems=yes|Castle Wars 2}}
{{WorldLine|355|United States|mems=yes|Blast Furnace}}
{{WorldLine|356|United States|mems=yes|Blast Furnace}}
{{WorldLine|357|United States|mems=yes|Blast Furnace}}
{{WorldLine|358|United Kingdom|mems=yes|Blast Furnace}}
{{WorldLine|359|Germany|mems=yes|-}}
{{WorldLine|360|Germany|mems=yes|-}}
{{WorldLine|361|United States|mems=yes|2000 skill total}}
{{WorldLine|362|United States|mems=yes|TzHaar Fight Pit}}
{{WorldLine|363|United Kingdom|mems=yes|2200 skill total}}
{{WorldLine|364|United Kingdom|mems=yes|1250 skill total}}
{{WorldLine|365|United Kingdom|mems=yes|High Risk World}}
{{WorldLine|366|United Kingdom|mems=yes|1500 skill total}}
{{WorldLine|367|Germany|mems=yes|-}}
{{WorldLine|368|Germany|mems=yes|-}}
{{WorldLine|369|United States|mems=yes|Wilderness PK - Members}}
{{WorldLine|370|United States|mems=yes|Fishing Trawler}}
{{WorldLine|371|United Kingdom|mems=yes|-}}
{{WorldLine|372|United Kingdom|mems=no|750 skill total}}
{{WorldLine|373|United Kingdom|mems=yes|1750 skill total}}
{{WorldLine|374|United States|mems=yes|Theatre of Blood}}
{{WorldLine|375|Germany|mems=yes|Zalcano}}
{{WorldLine|376|Germany|mems=yes|Theatre of Blood}}
{{WorldLine|377|United States|mems=yes|Mort'ton temple, Rat Pits}}
{{WorldLine|378|United States|mems=yes|Zalcano}}
{{WorldLine|379|United Kingdom|mems=no|PvP Arena (Legacy Duels)}}
{{WorldLine|380|United Kingdom|mems=no|-}}
{{WorldLine|381|United Kingdom|mems=no|500 skill total}}
{{WorldLine|382|United Kingdom|mems=no|-}}
{{WorldLine|383|Germany|mems=no|Castle Wars - Free}}
{{WorldLine|384|Germany|mems=no|-}}
{{WorldLine|385|United States|mems=yes|Brimhaven Agility}}
{{WorldLine|386|United States|mems=yes|Blast Furnace}}
{{WorldLine|387|Australia|mems=yes|Blast Furnace}}
{{WorldLine|388|Australia|mems=yes|Forestry}}
{{WorldLine|389|Australia|mems=yes|Wintertodt}}
{{WorldLine|390|Australia|mems=yes|LMS Competitive}}
{{WorldLine|391|Australia|mems=yes|1750 skill total}}
{{WorldLine|392|Australia|mems=yes|PvP World}}
{{WorldLine|393|United States|mems=no|750 skill total}}
{{WorldLine|394|United States|mems=yes|Yama}}
{{WorldLine|395|Germany|mems=yes|Blast Furnace}}
{{WorldLine|396|Germany|mems=yes|2000 skill total}}
{{WorldLine|397|Germany|mems=no|-}}
{{WorldLine|398|Germany|mems=no|Forestry}}
{{WorldLine|399|Germany|mems=no|-}}
{{WorldLine|401|United States|mems=yes|-}}
{{WorldLine|402|United States|mems=yes|Wintertodt}}
{{WorldLine|403|United States|mems=yes|-}}
{{WorldLine|404|United States|mems=yes|-}}
{{WorldLine|405|Germany|mems=yes|-}}
{{WorldLine|406|Germany|mems=yes|-}}
{{WorldLine|407|United Kingdom|mems=yes|-}}
{{WorldLine|408|United Kingdom|mems=yes|-}}
{{WorldLine|409|United States|mems=yes|Guardians of the Rift}}
{{WorldLine|410|United States|mems=yes|-}}
{{WorldLine|411|United States|mems=yes|-}}
{{WorldLine|412|Australia|mems=yes|-}}
{{WorldLine|413|Germany|mems=no|500 skill total}}
{{WorldLine|414|Germany|mems=no|750 skill total}}
{{WorldLine|415|United States|mems=yes|2200 skill total}}
{{WorldLine|416|United States|mems=yes|1500 skill total}}
{{WorldLine|417|United States|mems=no|-}}
{{WorldLine|418|United States|mems=no|-}}
{{WorldLine|419|United States|mems=no|500 skill total}}
{{WorldLine|420|United States|mems=yes|1500 skill total}}
{{WorldLine|421|United States|mems=yes|-}}
{{WorldLine|422|United States|mems=yes|Tempoross}}
{{WorldLine|423|United States|mems=yes|Fresh Start}}
{{WorldLine|424|Australia|mems=yes|Blast Furnace}}
{{WorldLine|425|Australia|mems=yes|-}}
{{WorldLine|426|Australia|mems=yes|House Party, Gilded Altar}}
{{WorldLine|427|Australia|mems=no|500 skill total}}
{{WorldLine|428|United States|mems=yes|2000 skill total}}
{{WorldLine|429|United States|mems=yes|1250 skill total}}
{{WorldLine|430|United States|mems=no|-}}
{{WorldLine|431|United States|mems=no|-}}
{{WorldLine|432|United States|mems=no|750 skill total}}
{{WorldLine|433|United States|mems=no|-}}
{{WorldLine|434|United States|mems=no|Forestry}}
{{WorldLine|435|United States|mems=no|-}}
{{WorldLine|436|United States|mems=no|-}}
{{WorldLine|437|United States|mems=no|-}}
{{WorldLine|438|United States|mems=yes|-}}
{{WorldLine|439|United States|mems=yes|-}}
{{WorldLine|440|United States|mems=yes|-}}
{{WorldLine|441|United States|mems=yes|-}}
{{WorldLine|442|United States|mems=yes|-}}
{{WorldLine|443|United States|mems=yes|-}}
{{WorldLine|444|United States|mems=yes|Forestry}}
{{WorldLine|445|United States|mems=yes|Guardians of the Rift}}
{{WorldLine|446|United States|mems=yes|Role-playing}}
{{WorldLine|447|Germany|mems=yes|1250 skill total}}
{{WorldLine|448|Germany|mems=yes|1500 skill total}}
{{WorldLine|449|Germany|mems=yes|1750 skill total}}
{{WorldLine|450|Germany|mems=yes|2200 skill total}}
{{WorldLine|451|Germany|mems=no|-}}
{{WorldLine|452|Germany|mems=no|-}}
{{WorldLine|453|Germany|mems=no|-}}
{{WorldLine|454|Germany|mems=no|-}}
{{WorldLine|455|Germany|mems=no|-}}
{{WorldLine|456|Germany|mems=no|-}}
{{WorldLine|457|Germany|mems=yes|-}}
{{WorldLine|458|Germany|mems=yes|-}}
{{WorldLine|459|Germany|mems=yes|-}}
{{WorldLine|461|Germany|mems=yes|-}}
{{WorldLine|462|Germany|mems=yes|-}}
{{WorldLine|463|Germany|mems=yes|Tempoross}}
{{WorldLine|464|Germany|mems=yes|Guardians of the Rift}}
{{WorldLine|465|Germany|mems=yes|House Party, Gilded Altar}}
{{WorldLine|466|Germany|mems=yes|Blast Furnace}}
{{WorldLine|467|United States|mems=yes|1750 skill total}}
{{WorldLine|468|United States|mems=no|500 skill total}}
{{WorldLine|469|United States|mems=no|LMS Casual}}
{{WorldLine|470|United States|mems=yes|-}}
{{WorldLine|471|United States|mems=yes|-}}
{{WorldLine|472|United States|mems=yes|-}}
{{WorldLine|473|United States|mems=yes|-}}
{{WorldLine|474|United States|mems=yes|High Risk World}}
{{WorldLine|475|United States|mems=yes|-}}
{{WorldLine|476|United States|mems=yes|-}}
{{WorldLine|477|United States|mems=yes|Clan Recruitment}}
{{WorldLine|478|United States|mems=yes|Guardians of the Rift}}
{{WorldLine|479|United States|mems=yes|PvP Arena (Legacy Duels)}}
{{WorldLine|480|United States|mems=yes|Ourania Altar}}
{{WorldLine|481|United States|mems=yes|Nex FFA}}
{{WorldLine|482|United States|mems=yes|-}}
{{WorldLine|483|United States|mems=no|PvP Arena (Legacy Duels)}}
{{WorldLine|484|United States|mems=yes|-}}
{{WorldLine|485|United States|mems=yes|-}}
{{WorldLine|486|United States|mems=yes|-}}
{{WorldLine|487|United States|mems=yes|Forestry}}
{{WorldLine|488|United States|mems=yes|-}}
{{WorldLine|489|United States|mems=yes|-}}
{{WorldLine|490|United States|mems=yes|-}}
{{WorldLine|491|United States|mems=yes|Burthorpe Games Room}}
{{WorldLine|492|United States|mems=yes|-}}
{{WorldLine|493|United States|mems=yes|Pyramid Plunder}}
{{WorldLine|494|United States|mems=yes|Blast Furnace}}
{{WorldLine|495|United States|mems=yes|Blast Furnace}}
{{WorldLine|496|United States|mems=yes|Blast Furnace}}
{{WorldLine|497|United Kingdom|mems=no|Clan Recruitment}}
{{WorldLine|498|United Kingdom|mems=no|-}}
{{WorldLine|499|United Kingdom|mems=no|-}}
{{WorldLine|500|United Kingdom|mems=yes|-}}
{{WorldLine|501|United Kingdom|mems=yes|-}}
{{WorldLine|502|United Kingdom|mems=yes|Speedrunning World}}
{{WorldLine|503|United Kingdom|mems=yes|-}}
{{WorldLine|504|United Kingdom|mems=yes|-}}
{{WorldLine|505|United Kingdom|mems=yes|Nex FFA}}
{{WorldLine|506|United Kingdom|mems=yes|PvP Arena (Legacy Duels)}}
{{WorldLine|507|United Kingdom|mems=yes|-}}
{{WorldLine|508|United Kingdom|mems=yes|Brimhaven Agility}}
{{WorldLine|509|United Kingdom|mems=yes|-}}
{{WorldLine|510|United Kingdom|mems=yes|Forestry}}
{{WorldLine|511|United Kingdom|mems=yes|-}}
{{WorldLine|512|United Kingdom|mems=yes|House Party, Gilded Altar}}
{{WorldLine|513|United Kingdom|mems=yes|Zeah Runecrafting}}
{{WorldLine|514|United Kingdom|mems=yes|Nightmare of Ashihama}}
{{WorldLine|515|United Kingdom|mems=yes|Blast Furnace}}
{{WorldLine|516|United Kingdom|mems=yes|Blast Furnace}}
{{WorldLine|517|United Kingdom|mems=yes|Varlamore PvM}}
{{WorldLine|518|United Kingdom|mems=yes|-}}
{{WorldLine|519|United Kingdom|mems=yes|-}}
{{WorldLine|520|United Kingdom|mems=yes|-}}
{{WorldLine|521|United Kingdom|mems=yes|-}}
{{WorldLine|522|United Kingdom|mems=yes|Guardians of the Rift}}
{{WorldLine|523|United Kingdom|mems=yes|Nex FFA}}
{{WorldLine|524|United Kingdom|mems=yes|-}}
{{WorldLine|525|United Kingdom|mems=yes|-}}
{{WorldLine|526|Australia|mems=yes|2200 skill total}}
{{WorldLine|527|Australia|mems=yes|2000 skill total}}
{{WorldLine|528|Australia|mems=yes|1500 skill total}}
{{WorldLine|529|Australia|mems=yes|1250 skill total}}
{{WorldLine|530|Australia|mems=no|750 skill total}}
{{WorldLine|531|Australia|mems=yes|Tombs of Amascut}}
{{WorldLine|532|Australia|mems=yes|Group PvM}}
{{WorldLine|533|Australia|mems=yes|High Risk World}}
{{WorldLine|534|Australia|mems=yes|Guardians of the Rift}}
{{WorldLine|535|Australia|mems=yes|Soul Wars}}
{{WorldLine|536|Australia|mems=yes|-}}
{{WorldLine|537|Australia|mems=no|-}}
{{WorldLine|538|United States|mems=yes|-}}
{{WorldLine|539|United States|mems=yes|PvP World}}
{{WorldLine|540|United States|mems=yes|Speedrunning World}}
{{WorldLine|541|United States|mems=yes|-}}
{{WorldLine|542|United States|mems=yes|-}}
{{WorldLine|543|United States|mems=yes|-}}
{{WorldLine|544|United States|mems=yes|-}}
{{WorldLine|545|United States|mems=yes|-}}
{{WorldLine|546|United States|mems=yes|-}}
{{WorldLine|547|United States|mems=yes|-}}
{{WorldLine|548|Germany|mems=yes|PvP World - High Risk}}
{{WorldLine|549|Germany|mems=yes|Speedrunning World}}
{{WorldLine|550|Germany|mems=yes|-}}
{{WorldLine|551|Germany|mems=yes|-}}
{{WorldLine|552|Germany|mems=no|-}}
{{WorldLine|553|Germany|mems=no|-}}
{{WorldLine|554|Germany|mems=no|-}}
{{WorldLine|555|Germany|mems=no|-}}
{{WorldLine|556|Germany|mems=yes|-}}
{{WorldLine|557|Germany|mems=yes|-}}
{{WorldLine|558|United Kingdom|mems=yes|PvP Arena (UK)}}
{{WorldLine|559|United Kingdom|mems=yes|LMS Competitive}}
{{WorldLine|562|United Kingdom|mems=yes|-}}
{{WorldLine|563|United Kingdom|mems=yes|-}}
{{WorldLine|564|United Kingdom|mems=yes|-}}
{{WorldLine|565|United Kingdom|mems=no|Fresh Start}}
{{WorldLine|566|United Kingdom|mems=yes|-}}
{{WorldLine|567|United Kingdom|mems=yes|Royal Titans}}
{{WorldLine|568|Australia|mems=yes|Speedrunning World}}
{{WorldLine|569|Australia|mems=yes|Bounty Hunter World}}
{{WorldLine|570|Australia|mems=yes|PvP Arena (AUS)}}
{{WorldLine|571|Australia|mems=no|PvP Arena (Legacy Duels)}}
{{WorldLine|573|United States|mems=yes|Royal Titans}}
{{WorldLine|574|United States|mems=yes|-}}
{{WorldLine|575|United States|mems=yes|-}}
{{WorldLine|576|United States|mems=yes|Speedrunning World}}
{{WorldLine|577|United States|mems=no|PvP World - Free}}
{{WorldLine|578|United States|mems=yes|PvP Arena (US)}}
{{WorldLine|581|United States|mems=yes|Event PvP World}}
{{WorldLine|582|Germany|mems=yes|-}}
{{WorldLine|590|Australia|mems=yes|-}}
{{WorldLine|591|Australia|mems=yes|-}}
{{WorldLine|592|Australia|mems=yes|-}}
{{WorldLine|593|Australia|mems=yes|-}}
{{WorldLine|594|Australia|mems=yes|-}}
{{WorldLine|595|Australia|mems=yes|-}}
{{WorldLine|598|United States|mems=yes|Locked Developer World}}
{{WorldLine|599|United States|mems=yes|-}}
{{WorldLine|600|United States|mems=yes|-}}
{{WorldLine|601|United States|mems=yes|-}}
{{WorldLine|602|United States|mems=yes|-}}
{{WorldLine|603|United States|mems=yes|-}}
{{WorldLine|604|United States|mems=yes|-}}
{{WorldLine|605|United States|mems=yes|-}}
{{WorldLine|606|United States|mems=yes|-}}
{{WorldLine|607|United States|mems=yes|-}}
{{WorldLine|608|United States|mems=yes|-}}
{{WorldLine|609|United States|mems=yes|-}}
{{WorldLine|610|United States|mems=yes|-}}
{{WorldLine|611|United States|mems=yes|-}}
{{WorldLine|612|United States|mems=yes|-}}
{{WorldLine|613|United States|mems=yes|-}}
{{WorldLine|614|United States|mems=yes|-}}
{{WorldLine|615|United States|mems=yes|-}}
{{WorldLine|616|United States|mems=yes|-}}
{{WorldLine|617|United States|mems=yes|-}}
{{WorldLine|618|United States|mems=yes|-}}
{{WorldLine|619|United Kingdom|mems=yes|-}}
{{WorldLine|620|United Kingdom|mems=yes|-}}
{{WorldLine|621|United Kingdom|mems=yes|-}}
{{WorldLine|622|United Kingdom|mems=yes|-}}
{{WorldLine|623|United Kingdom|mems=yes|-}}
{{WorldLine|624|Germany|mems=yes|-}}
{{WorldLine|625|Germany|mems=yes|-}}
{{WorldLine|626|Germany|mems=yes|-}}
{{WorldLine|627|Germany|mems=yes|-}}
{{WorldLine|628|Germany|mems=yes|-}}"
`;
